outlines/generate/generator.py

```diff
@@ -0,0 +1,359 @@
+import math
+from typing import TYPE_CHECKING, Callable, Iterator, List, Optional, Tuple, Union
+
+import torch
+
+from outlines.index.index import FSMState
+
+if TYPE_CHECKING:
+    from outlines.generate.samplers import Sampler
+    from outlines.index.index import FSM
+
+
+@dataclasses.dataclass(frozen=True)
+class Sequences:
+    next_tokens: Union[str, List[str]]
+    token_ids: torch.Tensor
+    kv_cache: torch.Tensor
+    logits: torch.Tensor
+
+
+class SequenceGenerator:
+    def __init__(self, fsm, model, sampler, device):
+        self.generate_token = token_generator(model, sampler)
+        self.fsm = fsm
+        self.tokenizer = model.tokenizer
+        self.device = device
+
+    def init_generation_state(
+        self,
+        prompt: Union[str, List[str]],
+        kv_cache: Optional[torch.Tensor] = None,
+    ):
+        """Initialize the generation state.
+
+        This method is responsible for encoding the prompt, moving token ids
+        to the device and initializing the random number generator.
+
+        Parameters
+        ----------
+        prompt
+            The prompt on which the generation is conditioned.
+        rng
+            The state of the random number generator.
+
+        Returns
+        -------
+        A `GenerationState` object.
+
+        """
+        token_ids, attention_masks = self.tokenizer.encode(prompt)
+        token_ids = token_ids.to(self.device)
+        attention_masks = attention_masks.to(self.device)
+
+        return token_ids, attention_masks, kv_cache
+
+    def format_sequence(self, sequence):
+        return sequence
+
+    def __call__(
+        self,
+        prompt,
+        kv_cache: Optional[torch.tensor] = None,
+        rng: Optional[torch.Generator] = None,
+    ) -> Union[str, List[str]]:
+        sequence_generator = self.stream(prompt, kv_cache, rng)
+        tokens = [token for token in sequence_generator]
+        sequences = [
+            self.format_sequence("".join(sequence)) for sequence in list(zip(*tokens))
+        ]
+        return sequences if len(sequences) > 1 else sequences[0]
+
+    def stream(
+        self,
+        prompt: str,
+        kv_cache: Optional[torch.tensor] = None,
+        rng: Optional[torch.Generator] = None,
+    ) -> Iterator[str]:
+        if rng is None:
+            rng = torch.Generator(device=self.device)
+            rng.seed()
+
+        init_state = self.init_generation_state(prompt, kv_cache)
+
+        token_ids = init_state[1]
+        num_sequences = token_ids.shape[0]
+        init_fsm_states = [FSMState(0) for _ in range(num_sequences)]
+
+        tokens = sequence_generator(
+            self.generate_token, self.fsm, init_state, init_fsm_states, rng
+        )
+
+        def token_generator():
+            while True:
+                try:
+                    sequence = next(tokens)
+                except StopIteration:
+                    return
+
+                next_tokens: str = sequence.next_tokens
```

Comment:
We do not return the logits and kv-cache because we would like to keep compatibility with the former API.

We may actually want to transition now while we have aliases.